### PR TITLE
Texture atlas rework

### PR DESCRIPTION
**Pull request by s-puig**
Adopted #5103 
_Not ready for review yet_

**Description _temporarily_ copied from #5103**

# Objective

> Old MR: https://github.com/bevyengine/bevy/pull/5072
> Associated UI MR: https://github.com/bevyengine/bevy/pull/5070
> Adresses https://github.com/bevyengine/bevy/issues/1618

- Unify sprite management

## Solution

- Remove the Handle<Image> field in TextureAtlas which is the main cause for all the boilerplate
- Remove the redundant TextureAtlasSprite component
- Renamed TextureAtlas asset to TextureAtlasLayout (https://github.com/bevyengine/bevy/pull/5103#discussion_r917281844)
- Add a TextureAtlas component, containing the atlas layout handle and the section index

The difference between this solution and https://github.com/bevyengine/bevy/pull/5072 is that instead of the enum approach is that we can more easily manipulate texture sheets without any breaking changes for classic SpriteBundles (@mockersf https://github.com/bevyengine/bevy/pull/5072#issuecomment-1165836139)

Also, this approach is more data oriented extracting the Handle<Image> and avoiding complex texture atlas manipulations to retrieve the texture in both applicative and engine code.
With this method, the only difference between a SpriteBundle and a SpriteSheetBundle is an additional component storing the atlas handle and the index.

This solution can be applied to bevy_ui as well (see https://github.com/bevyengine/bevy/pull/5070).

---

## Changelog

- (BREAKING) Removed TextureAtlasSprite
- (BREAKING) Renamed TextureAtlas to TextureAtlasLayout
- (BREAKING) SpriteSheetBundle:
  - Uses a Sprite instead of a TextureAtlasSprite component
  - Has a texture field containing a Handle<Image> like the SpriteBundle
  - Has a new TextureAtlas component instead of a Handle<TextureAtlasLayout>
- (BREAKING) DynamicTextureAtlasBuilder::add_texture takes an additional &Handle<Image> parameter
- (BREAKING) TextureAtlasLayout::from_grid no longer takes a Handle<Image> parameter
- (BREAKING) TextureAtlasBuilder::finish now returns a Result<(TextureAtlasLayout, Handle<Image>), _>
- bevy_text:
  - GlyphAtlasInfo stores the texture Handle<Image>
  - FontAtlas stores the texture Handle<Image>

## Migration Guide

```
fn my_system(
  mut images: ResMut<Assets<Image>>, 
-  mut atlases: ResMut<Assets<TextureAtlas>>, 
+  mut atlases: ResMut<Assets<TextureAtlasLayout>>, 
  asset_server: Res<AssetServer>
) {
    let texture_handle: asset_server.load("my_texture.png");
-   let layout = TextureAtlas::from_grid(texture_handle, Vec2::new(25.0, 25.0), 5, 5, None, None);
+   let layout = TextureAtlasLayout::from_grid(Vec2::new(25.0, 25.0), 5, 5, None, None);
    let layout_handle = atlases.add(atlas);
    commands.spawn_bundle(SpriteSheetBundle {
-      sprite: TextureAtlasSprite::new(0),
-      texture_atlas: atlas_handle,
+      atlas: TextureAtlas {
+         layout: layout_handle,
+         index: 0
+      },
+      texture: texture_handle,
       ..Default::default()
     });
}
```

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
cc @ManevilleF @mwbryant if you two have time and interest to participate in the discussion <3

**Comment by mwbryant** [comment]
> * [UI texture atlas support #8822](https://github.com/bevyengine/bevy/pull/8822) added atlas textures to UI. Should we merge the extract_atlas_nodes into a single system (extract_ui_nodes) like the [author suggested](https://github.com/bevyengine/bevy/pull/8822#issuecomment-1595037632) in this PR? Leave it to a second PR?

If you can, I would strongly recommend doing this in the same PR.  Keeping the UI and regular sprite usage in sync is important for consistency.  There shouldn't be anything stopping this and it would probably simplify some of the complexity/duplication I introduced :)

I can also handle it in a follow-up if you ping me and remind me.  

Thank you for taking this on by the way, unifying the two ways we handle sprites is very important in my opinion.  A lot of people start with single sprites for prototyping and then swap over to sheets and this will reduce friction there.

File: crates/bevy_sprite/src/texture_atlas.rs [hunk 1]
```diff
@@ -1,94 +1,85 @@
...
             size: dimensions,
             texture_handles: None,
             textures: Vec::new(),
         }
     }
 
-    /// Generate a [`TextureAtlas`] by splitting a texture into a grid where each
-    /// `tile_size` by `tile_size` grid-cell is one of the textures in the
+    /// Generate a [`TextureAtlasLayout`] as a grid where each
+    /// `tile_size` by `tile_size` grid-cell is one of *section* in the
```
Comment (mnmaita) [inline crates/bevy_sprite/src/texture_atlas.rs:64]:
```suggestion
    /// `tile_size` by `tile_size` grid-cell is one *section* in the
```
Minor language nit here.

**Review by mnmaita — COMMENTED** [review]
This is a great change, thanks for adopting this PR!

**Comment by ManevilleF** [comment]
Hi, thanks for adopting my old PR. About your questions:

- I think you should handle both UI and sprites in the same PR, I originally opened two separate pull requests (#5103 and #5070) because the functionality did not exist at all UI side

- About splitting the `TextureAtlas` component, there was a discussion and we decided to use a single component as it wouldn't make sense to split it and have a single index component (https://github.com/bevyengine/bevy/pull/5070#discussion_r903560834 https://github.com/bevyengine/bevy/pull/5103#discussion_r917281844) and it might make the API/Bundle less clear but the situation may have changed

File: examples/2d/sprite_sheet.rs [hunk 2]
```diff
@@ -43,19 +39,21 @@ fn animate_sprite(
...
-    mut texture_atlases: ResMut<Assets<TextureAtlas>>,
+    mut texture_atlases: ResMut<Assets<TextureAtlasLayout>>,
 ) {
-    let texture_handle = asset_server.load("textures/rpg/chars/gabe/gabe-idle-run.png");
-    let texture_atlas =
-        TextureAtlas::from_grid(texture_handle, Vec2::new(24.0, 24.0), 7, 1, None, None);
-    let texture_atlas_handle = texture_atlases.add(texture_atlas);
+    let texture = asset_server.load("textures/rpg/chars/gabe/gabe-idle-run.png");
+    let atlas = TextureAtlasLayout::from_grid(Vec2::new(24.0, 24.0), 7, 1, None, None);
+    let texture_atlas = texture_atlases.add(atlas);
```
Comment (ManevilleF) [inline examples/2d/sprite_sheet.rs:46]: Maybe the two variables should be renamed `atlas_layout` and `layout_handle` for clarity

**Comment by alice-i-cecile** [comment]
Closing out due to inactivity for now. Feel free to reopen if you find you have the time to tackle this.

**Comment by alice-i-cecile** [comment]
Sounds good, reopen this when you do :)

**Pull request closed without merging**